src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -197,20 +197,41 @@ private void AllocateWriteHeadSynchronized(int sizeHint)
                             _writingHeadBytesBuffered = 0;
                         }
 
-                        BufferSegment newSegment = AllocateSegment(sizeHint);
+                        if (_writingHead.Length == 0)
```

Comment:
I would use _writingHeadBytesBuffered instead (but the logic would need to be reworked a bit)

Reply:
I think I tried using that field initially, but it gets set to 0 in Advance with a non-zero value, so you need to check some other value anyways. What's wrong with `_writingHead.Length`?

Reply:
Nothing wrong with it, it just avoids the indirection but this is fine.